Fix rustdoc warnings and test cargo doc in CI

arpad-m opened the pull request:
## Problem

`cargo +nightly doc` is giving a lot of warnings: broken links, naked URLs, etc.

## Summary of changes

* update the `proc-macro2` dependency so that it can compile on latest Rust nightly, see https://github.com/dtolnay/proc-macro2/pull/391 and https://github.com/dtolnay/proc-macro2/issues/398
* allow the `private_intra_doc_links` lint, as linking to something that's private is always more useful than just mentioning it without a link: if the link breaks in the future, at least there is a warning due to that. Also, one might enable [`--document-private-items`](https://doc.rust-lang.org/cargo/commands/cargo-doc.html#documentation-options) in the future and make these links work in general.
* fix all the remaining warnings given by `cargo +nightly doc`
* make it possible to run `cargo doc` on stable Rust by updating `opentelemetry` and associated crates to version 0.19, pulling in a fix that previously broke `cargo doc` on stable: https://github.com/open-telemetry/opentelemetry-rust/pull/904
* Add `cargo doc` to CI to ensure that it won't get broken in the future. 

Fixes #2557

## Future work
* Potentially, it might make sense, for development purposes, to publish the generated rustdocs somewhere, like for example [how the rust compiler does it](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_driver/index.html). I will file an issue for discussion.

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

vadim2404 commented:
Should we incorporate this check into CI?

koivunej commented on libs/pageserver_api/src/models.rs:
```diff
@@ -411,12 +411,16 @@ pub struct LayerResidenceEvent {
     /// The layer map is being populated, e.g. during timeline load or attach.
     /// This includes [`RemoteLayer`] objects created in [`reconcile_with_remote`].
     /// We need to record such events because there is no persistent storage for the events.
+    ///
+    // https://github.com/rust-lang/rust/issues/74481
+    /// [`RemoteLayer`]: ../pageserver/tenant/storage_layer/struct.RemoteLayer.html
+    /// [`reconcile_with_remote`]: ../pageserver/tenant/struct.Timeline.html#method.reconcile_with_remote
```
Is the rust issue meant to be linekd here?

arpad-m replied:
The rust issue is just that we have to link to the html because intra doc links don't support linking to dependents.

But the links actually don't work, I need to fix them, thanks for bringing this up.

koivunej reviewed (approved):
I think these are good changes.

vadim2404 commented:
related to #2557

arpad-m commented:
> Should we incorporate this check into CI?

I think so. Originally this PR didn't have such a CI check as nightly was required to run `cargo doc` and CI currently only has one stable Rust version. But then I changed the PR to update `opentelemetry` to 0.19 to fix the stable `cargo doc` build, so I've now added it to CI.

arpad-m commented:
Okay, a trial edit to add a broken link to the code [is showing a failure as expected](https://github.com/neondatabase/neon/actions/runs/5549304618/jobs/10133245878?pr=4711), very nice. I've removed that edit from the CI now.

koivunej reviewed (approved):
Still looking good.

I don't think we have any tests for the otel stuff, the parts that we use. They don't publish a clear changelog so ... aah they have per crate changelog:

- http 0.7 -> 0.8, looks quite boring
- otlp 0.11 -> 0.12, similarly boring, deps (within and outside otel)
- reqwest-tracing bump, for otel
- otel semantic conventions, deps
- proc_macro2 was for latest nightlies I understood?

I think these are all looking very mundane version bumps, and thus we could merge this even today but perhaps it's best to still consider if we release and what we release. So let's hold merging until that has been clarified.

koivunej commented on .github/workflows/build_and_test.yml:
```diff
@@ -127,6 +127,11 @@ jobs:
         run: cargo hack --feature-powerset clippy --release $CLIPPY_COMMON_ARGS
 
+      - name: Check documentation generation
+        run: cargo doc --all
```
Bubbling these up from DM's:

- `--all` => `--workspace` (as I am always confused by `--all` -- though, since we have top level crate without code it probably defaults to that)
- how about `--no-deps`?
- does `--document-private-items` here affect any lints or will we get the same regardless?

koivunej replied:
```suggestion
        run: cargo doc --workspace
```

arpad-m replied:
>     * how about `--no-deps`? While this PR does upgrade some deps to get them to be `cargo doc`'d I've just always used `--no-deps` for the less walltime spent; is there nowdays some loss of functionality I am not aware of (assuming we'll publish these in future)?

Good point

>     * would a `--document-private-items` here affect any lints or will we get the same regardless?

I think there won't be any change. I think it's cleaner to pass that instead of passing `-Arustdoc::private_intra_doc_links`.

koivunej replied:
Should we add a `--document-private-items` to this cmdline?

arpad-m replied:
I've added it.
